File: deltafi-api/lib/deltafi/monitor/status/checks/ingress_status_check.rb
#
#    DeltaFi - Data transformation and enrichment platform
#
#    Copyright 2021-2023 DeltaFi Contributors <deltafi@deltafi.org>
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#

# frozen_string_literal: true

require 'deltafi/common'
require 'deltafi/monitor/status/check'
require 'bytesize'

module Deltafi
  module Monitor
    module Status
      module Checks
        class IngressStatusCheck < Status::Check
          DEFAULT_REQUIRED_MEGABYTES = 1
          REQUIRED_MEGABYTES_PROPERTY = %w[ingressDiskSpaceRequirementInMb].freeze
          INGRESS_ENABLED_PROPERTY = %w[ingressEnabled].freeze
          FLOW_ERRORS_QUERY = 'query { dataSourceErrorsExceeded { name maxErrors currErrors } }'

          @@ingress_disabled_by_storage = false
          @@disabled_flows = []

          def initialize
            super('Ingress Status Check')
          end

          def run
            @required_bytes = required_bytes

            check_for_disabled_ingress
            check_for_storage_disabled_ingress
            check_for_disabled_flows

            self
          end

          private

          def check_for_disabled_ingress
            return if ingress_enabled?

            self.code = 1
            message_lines << "##### Ingress is disabled\n"
            message_lines << "Reenable the system property '#{INGRESS_ENABLED_PROPERTY}' to restart ingress."
          end

          def format_number(number)
            number.to_i.to_s.gsub(/\B(?=(\d{3})*\b)/, ',')
          end

          def check_for_storage_disabled_ingress
            remaining = remaining_bytes
            required = required_bytes

            return unless remaining && required

            if remaining >= required
              notify_storage_enabled
            else
              notify_storage_disabled remaining, required
              self.code = 1
              message_lines << "##### Ingress is disabled due to lack of content storage\n"
              message_lines << "Required bytes in content storage: #{format_number(required)} (#{ByteSize.new(required)})\n"
              message_lines << "Remaining bytes in content storage: #{format_number(remaining)} (#{ByteSize.new(remaining)})\n"
            end
          end

          def notify_storage_disabled(remaining, required)
            return if @@ingress_disabled_by_storage

            content_lines = [
              "- Remaining bytes in content storage: #{format_number(remaining)} (#{ByteSize.new(remaining)})",
              "- Required bytes: #{format_number(required)} (#{ByteSize.new(required)})"
            ]
            Deltafi::Events.generate 'Disabling ingress due to depleted content storage',
                                     content: content_lines.join("\n"),
                                     severity: 'warn',
                                     notification: true,
                                     source: 'ingress'
            @@ingress_disabled_by_storage = true
          end

          def notify_storage_enabled
            return unless @@ingress_disabled_by_storage

            Deltafi::Events.generate 'Ingress is re-enabled',
                                     severity: 'info',
                                     notification: true,
                                     source: 'ingress'
            @@ingress_disabled_by_storage = false
          end

          def required_bytes
            DF::SystemProperties.dig(REQUIRED_MEGABYTES_PROPERTY, DEFAULT_REQUIRED_MEGABYTES).to_i * 1_000_000
          end

          def remaining_bytes
            json = DF::API::V1::Metrics::System.content
            limit = json&.dig(:limit)&.to_i
            usage = json&.dig(:usage)&.to_i
            return unless limit && usage

            limit - usage
          end

          def ingress_enabled?
            DF::SystemProperties.dig(INGRESS_ENABLED_PROPERTY, 'true').to_s.casecmp('true').zero?
          end

          def check_for_disabled_flows
            disabled = current_disabled_flows

            (disabled.keys - @@disabled_flows).each do |error_flow|
              notify_flow_disabled(error_flow, disabled[error_flow][:currErrors], disabled[error_flow][:maxErrors])
            end

            (@@disabled_flows - disabled.keys).each do |fixed_flow|
              notify_flow_reenabled(fixed_flow)
            end

            @@disabled_flows = disabled.keys

            return if @@disabled_flows.empty?

            self.code = 1
            message_lines << "##### Ingress is disabled for flows with too many errors\n"
            message_lines << 'Acknowledge or resolve errors on these flows to continue:'
            disabled.each do |k, v|
              message_lines << "\n- #{k}: #{v[:currErrors]} errors, #{v[:maxErrors]} allowed"
            end
          end

          def notify_flow_disabled(flow, curr_errors, max_errors)
            Deltafi::Events.generate "Alert: Disabling ingress to flow #{flow} due to too many errors",
                                     content: "- Current errors for flow: #{curr_errors}\n- Maximum errors allowed: #{max_errors}\n",
                                     severity: 'warn',
                                     notification: true,
                                     source: 'ingress'
          end

          def notify_flow_reenabled(flow)
            Deltafi::Events.generate "Ingress is re-enabled for flow #{flow}",
                                     severity: 'info',
                                     notification: true,
                                     source: 'ingress'
          end

          def current_disabled_flows
            response = DF.graphql(FLOW_ERRORS_QUERY)
            parsed_response = JSON.parse(response.body, symbolize_names: true)
            raise StandardError, parsed_response[:errors]&.first&.dig(:message) if parsed_response.key?(:errors)

            flow_errors = parsed_response.dig(:data, :dataSourceErrorsExceeded) || []
            flow_errors.to_h do |r|
              [ r[:name], { maxErrors: r[:maxErrors], currErrors: r[:currErrors] } ]
            end
          end
        end
      end
    end
  end
end

File: deltafi-core/src/main/java/org/deltafi/core/monitor/checks/IngressStatusCheck.java
/*
 *    DeltaFi - Data transformation and enrichment platform
 *
 *    Copyright 2021-2025 DeltaFi Contributors <deltafi@deltafi.org>
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package org.deltafi.core.monitor.checks;

import lombok.extern.slf4j.Slf4j;
import org.deltafi.core.exceptions.StorageCheckException;
import org.deltafi.core.generated.types.DataSourceErrorState;
import org.deltafi.core.monitor.MonitorProfile;
import org.deltafi.core.monitor.checks.CheckResult.ResultBuilder;
import org.deltafi.core.services.*;
import org.deltafi.core.types.Event;
import org.deltafi.core.types.Event.Severity;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@MonitorProfile
@Slf4j
public class IngressStatusCheck extends StatusCheck {

    private final EventService eventService;
    private final SystemService systemService;
    private final RestDataSourceService restDataSourceService;
    private final TimedDataSourceService timedDataSourceService;
    private final DeltaFiPropertiesService propertiesService;

    private boolean ingressDisabledByStorage = false;
    private Set<String> disabledFlows = new HashSet<>();

    public IngressStatusCheck(EventService eventService, SystemService systemService,
                              RestDataSourceService restDataSourceService, TimedDataSourceService timedDataSourceService,
                              DeltaFiPropertiesService deltaFiPropertiesService) {
        super("Ingress Status Check");
        this.eventService = eventService;
        this.systemService = systemService;
        this.restDataSourceService = restDataSourceService;
        this.timedDataSourceService = timedDataSourceService;
        this.propertiesService = deltaFiPropertiesService;
    }

    @Override
    public CheckResult check() {
        ResultBuilder resultBuilder = new ResultBuilder();
        checkForDisabledIngress(resultBuilder);
        checkForStorageDisabledIngress(resultBuilder);
        checkForDisabledFlows(resultBuilder);
        return result(resultBuilder);
    }

    private void checkForDisabledIngress(ResultBuilder resultBuilder) {
        if (propertiesService.getDeltaFiProperties().isIngressEnabled()) {
            return;
        }

        resultBuilder.code(1);
        resultBuilder.addHeader("Ingress is disabled");
        resultBuilder.addLine("Reenable the system property 'ingressEnabled' to restart ingress.");
    }

    private void checkForStorageDisabledIngress(ResultBuilder resultBuilder) {
        try {
            doCheckForStorageDisabledIngress(resultBuilder);
        } catch (StorageCheckException e) {
            resultBuilder.code(1);
            resultBuilder.addHeader("Failed to get storage usage information");
            resultBuilder.addLine(e.getMessage());
        }
    }

    private void doCheckForStorageDisabledIngress(ResultBuilder resultBuilder) throws StorageCheckException {
        long remaining = systemService.contentNodeDiskMetrics().bytesRemaining();
        long required = propertiesService.getDeltaFiProperties().getIngressDiskSpaceRequirementInBytes();

        if (remaining > required) {
            notifyStorageEnabled();
        } else {
            Bytes remainingBytes = Bytes.bytes(remaining);
            Bytes requiredBytes = Bytes.bytes(required);
            notifyStorageDisabled(remainingBytes, requiredBytes);
            resultBuilder.code(1);
            resultBuilder.addHeader("Ingress is disabled due to lack of content storage");
            resultBuilder.addLine("Required bytes in content storage: " + requiredBytes.fullString + " (" + requiredBytes.humanReadable + ")\n");
            resultBuilder.addLine("Remaining bytes in content storage: " + remainingBytes.fullString + " (" + remainingBytes.humanReadable + ")\n");
        }
    }

    private void checkForDisabledFlows(ResultBuilder resultBuilder) {
        List<DataSourceErrorState> errorsExceededFlows = new ArrayList<>(restDataSourceService.dataSourceErrorsExceeded());
        errorsExceededFlows.addAll(timedDataSourceService.dataSourceErrorsExceeded());
        Set<String> errorsExceededFlowNames = errorsExceededFlows.stream()
                .map(DataSourceErrorState::getName).collect(Collectors.toSet());

        for (DataSourceErrorState disabledFlow : errorsExceededFlows) {
            if (!this.disabledFlows.contains(disabledFlow.getName())) {
                notifyFlowDisabled(disabledFlow);
            }
        }

        for (String maybeHealth : this.disabledFlows) {
            if (!errorsExceededFlowNames.contains(maybeHealth)) {
                notifyFlowReenabled(maybeHealth);
            }
        }

        this.disabledFlows = errorsExceededFlowNames;

        if (errorsExceededFlowNames.isEmpty()) {
            return;
        }

        resultBuilder.code(1);
        resultBuilder.addHeader("Ingress is disabled for flows with too many errors");
        resultBuilder.addLine("Acknowledge or resolve errors on these flows to continue:");
        for (DataSourceErrorState disabledFlow : errorsExceededFlows) {
            resultBuilder.addLine("\n- " + disabledFlow.getName() + ": " + disabledFlow.getCurrErrors()
                    + " errors, " + disabledFlow.getMaxErrors() + " allowed");
        }
    }

    private void notifyStorageEnabled() {
        if (!ingressDisabledByStorage) {
            return;
        }

        info("Ingress is re-enabled");
        ingressDisabledByStorage = false;
    }

    private void notifyStorageDisabled(Bytes remaining, Bytes required) {
        if (ingressDisabledByStorage) {
            return;
        }
        String contentLines = "- Remaining bytes in content storage: " + remaining.fullString + " (" + remaining.humanReadable + ")\n" +
                "- Remaining bytes in content storage: " + required.fullString + " (" + required.humanReadable + ")";
        warn("Disabling ingress due to depleted content storage", contentLines);
        ingressDisabledByStorage = true;
    }

    public void notifyFlowDisabled(DataSourceErrorState error) {
        String summary = "Alert: Disabling ingress to flow " + error.getName() + " due to too many errors";
        String content = "- Current errors for flow: " + error.getCurrErrors() +
                "\\n- Maximum errors allowed: " + error.getMaxErrors() + "\\n";

        warn(summary, content);
    }

    public void notifyFlowReenabled(String flow) {
        info("Ingress is re-enabled for flow " + flow);
    }

    private void info(String summary) {
        createEvent(summary, null, Severity.INFO);
    }


    private void warn(String summary, String content) {
        createEvent(summary, content, Severity.WARN);
    }

    private void createEvent(String summary, String content, String severity) {
        eventService.createEvent(Event.builder()
                .summary(summary)
                .content(content)
                .severity(severity)
                .notification(true)
                .source("ingress").build());
    }

    private record Bytes(String fullString, String humanReadable) {
        public static Bytes bytes(long value) {
            return new Bytes(prettyLong(value), humanLong(value));
        }

        private static String prettyLong(long value) {
            return "%,d".formatted(value);
        }

        private static String humanLong(long bytes) {
            if (bytes < 1000) return bytes + " B";
            int exp = (int) (Math.log(bytes) / Math.log(1000));
            String pre = ("kMGTPE").charAt(exp - 1) + "";
            return String.format("%.1f %sB", bytes / Math.pow(1000, exp), pre);
        }
    }
}
